### PR TITLE
[Datasets] Add `from_tf`

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji <balaji@anyscale.com>

**Blocked on**
- [x] #29590 

## Why are these changes needed?

Parity with `from_huggingface`. Also, `SimpleTensorflowDatasource` is clunky. For more motivation, read #29430.

## Related issue number

Towards #29430. See also #29589.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/read_api.py [hunk 1]
```diff
@@ -1326,6 +1327,58 @@ def convert(ds: "datasets.Dataset") -> Dataset[ArrowRow]:
         )
 
 
+@PublicAPI
+def from_tf(
```
Comment (matthewdeng) [inline python/ray/data/read_api.py:1331]: Can we expand this? 
```suggestion
def from_tensorflow(
```

Comment (bveeramani) [reply python/ray/data/read_api.py:1331]: We have a method called `Dataset.to_tf`. I prefer `tensorflow` over `tf`, but we should be consistent. 

What do you think? We could always rename `Dataset.to_tf` to `Dataset.to_tensorflow`.

Comment (matthewdeng) [reply python/ray/data/read_api.py:1331]: Ah right... Maybe we can:
1. Name this `from_tensorflow`.
2. Create a `to_tensorflow`.
3. Have `to_tf` redirect to `to_tensorflow` and eventually deprecate it.

But will leave to folks like @clarkzinzow @c21 to make the call here.

Comment (clarkzinzow) [reply python/ray/data/read_api.py:1331]: I'd actually vote for renaming these to `from_tf_dataset`, `to_tf_dataset`, `from_torch_dataset`, `to_torch_dataset`, etc. This is more explicit, discoverable, and better matches what the rest of the ecosystem is doing, e.g. HuggingFace: https://huggingface.co/docs/datasets/v2.6.1/en/package_reference/main_classes#datasets.Dataset.to_tf_dataset

I think that `tf` is a common enough shorthand that we shouldn't expand it to `tensorflow`, but happy to hear others' opinions, and maybe we can do a quick ecosystem check to verify this.

File: python/ray/data/read_api.py [hunk 2]
```diff
@@ -1326,6 +1327,58 @@ def convert(ds: "datasets.Dataset") -> Dataset[ArrowRow]:
...
+        [219, 185, 155],
+        [221, 187, 157]]], dtype=uint8), 'label': 7}]
+
+    Args:
+        dataset: A TensorFlow dataset.
+
+    Returns:
+        A :class:`Dataset` that contains the samples stored in the TensorFlow dataset.
+    """  # noqa: E501
+    return from_items(list(dataset.as_numpy_iterator()))
```
Comment (clarkzinzow) [inline python/ray/data/read_api.py:1379]: Misc. thing to note that I hit before in tests: `.as_numpy_iterator()` doesn't work with ragged tensors. https://github.com/tensorflow/tensorflow/issues/53149

**Review by clarkzinzow — APPROVED** [review]
LGTM

**Review by richardliaw — APPROVED** [review]

**Pull request merged**